Add continuous integration pipeline using GitHub actions

adamreeve opened the pull request:
This adds a CI pipeline using GitHub actions that runs on all pushes to master and for any pull requests. The jobs will build with Debug and Release configurations and run the CPU tests, then create a NuGet package artifact.

In order to make this work I've added targets to the csproj files that allow building all T4 templates automatically, although I haven't made these run as part of the normal build as they make it take a lot longer.

You can see an example run on my fork: https://github.com/adamreeve/ILGPU/actions/runs/110085962

~~This only runs on Windows currently due to the tests failing on Linux (#116).~~ I've added Linux tests but have skipped the problematic tests when running under Linux for now.

m4rs-mt commented:
@adamreeve A great job, thank you very much for submitting the PR and adding CI build features to this project.

m4rs-mt commented on Src/ILGPU.Tests.CPU/ILGPU.Tests.CPU.csproj:
```diff
@@ -34,19 +39,16 @@
-
-  <ItemGroup>
-    <None Update="Configurations.tt">
+    <None Remove="Configurations.tt" />
```
I wonder about these `Remove` and `Include` operations. I suppose this is required to create the T4 templates automatically, right?

adamreeve replied:
Ah so this was needed when I first used Microsoft.TextTemplating.targets to run the transforms as that only worked with Content items, but now I've switched to using dotnet-t4 (because the first approach doesn't work with the dotnet commands) I realise I can revert this change and keep these as None items, so I'll do that to keep things simpler.

The tt files are automatically included as "None" items so this was removing them from the None items and then adding them as Content items.

m4rs-mt commented on Src/ILGPU.Tests.Cuda/ILGPU.Tests.Cuda.csproj:
```diff
@@ -66,4 +68,11 @@
+      <TextTemplate Include="@(Content)" Condition="'%(Content.Extension)' == '.tt'">
+    </TextTemplate>
+  </ItemGroup>
+    <Exec WorkingDirectory="$(ProjectDir)" Command="dotnet tool run t4 %(TextTemplate.Identity) -I $([MSBuild]::ValueOrDefault('%(TextTemplate.RelativeDir)', '.')) -o %(TextTemplate.RelativeDir)%(TextTemplate.Filename).cs" />
```
Nice extension to automatically transform all T4 templates 🤓

m4rs-mt commented on Src/ILGPU/Frontend/Intrinsic/RemappedIntrinsics.Generated.tt:
```diff
@@ -96,16 +96,16 @@ namespace ILGPU.Frontend.Intrinsic
 
-<# foreach (var (method, type, _, className) in #>
-<#      unaryMathFunctions.Where(t => t.Item2 != null)) { #>
+<# foreach (var (method, type, _, className) in unaryMathFunctions
```
This is interesting; it looks like the T4-based parser may fail when using the previous code.

adamreeve replied:
Yes the template wouldn't build for me without this change, I wasn't sure whether that was due to a difference in Visual Studio versions or whether it was broken for everyone but hadn't been noticed.

m4rs-mt replied:
I've spend some time investigating these issues. It seems to be a problem with different SDK versions.

m4rs-mt replied:
I have encountered similar problems in the past regarding the "linefeed" characters appended to the end of the file, which had to be removed to have build support for Linux.

GPSnoopy commented on Src/ILGPU.Tests.OpenCL/ILGPU.Tests.OpenCL.csproj:
```diff
@@ -66,4 +68,11 @@
     </Compile>
   </ItemGroup>
 
+  <Target Name="TextTemplateTransform">
```
It's probably worth consolidating all of these behind a single target file that you can import inside your csproj.

jgiannuzzi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+      - name: Upload bin directory
+        uses: actions/upload-artifact@v2
+        with:
+          name: ${{ runner.os }}-bin
```
Is templating with `${{ runner.os }}` necessary?
I guess the idea is to have this job run on multiple platforms once possible, but we would only upload the Windows build, right?

jgiannuzzi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+        working-directory: ./Src
+      - name: CPU Tests
+        run: dotnet test --configuration=Release
+        working-directory: ./Src/ILGPU.Tests.CPU
```
Can we use `dotnet test --configuration=Release Src/ILGPU.Tests.CPU` instead of setting `working-directory`? Same remark for all the other `working-directory` instances except for `dotnet tool restore`.

jgiannuzzi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+        with:
+          name: Windows-bin
+          path: Bin
+      - name: Set up NuGet
+        uses: nuget/setup-nuget@v1
+        with:
+          nuget-version: '5.x'
+      - name: Create NuGet package
+        run: nuget pack ILGPU.nuspec
+        working-directory: .nuget
```
I wonder whether we should not create the nuget package directly inside the previous job. In that case, there would be no need to upload the bin directory as an artifact. The resulting artifact would only be the nuget package.

In the future, once we can run tests on multiple platforms, we might want to have a separate job for publishing the nuget package when all tests succeed and the CI is running for a tag like `v*`.

m4rs-mt replied:
Sounds reasonable; furthermore, this should simplify the whole internal deployment/test process.

adamreeve replied:
Yeah I separated them out thinking it would be cleaner if we built on multiple OSs and then created a NuGet package just from the Windows build, but we could still do that by conditionally running the NuGet package step and it would be simpler.

jgiannuzzi commented:
Awesome job, @adamreeve!

adamreeve commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,59 @@
+        # TargetFramework specified so this only runs once rather than per framework
+        run: dotnet msbuild -t:TextTemplateTransform -p:TargetFramework=NA
+        working-directory: ./Src
+      - name: Debug Build
+        run: dotnet build --configuration=Debug -p:TreatWarningsAsErrors=true
+        working-directory: ./Src
```
:+1: That would simplify things, we could then use dotnet pack to build a package straight from the project and not need to install nuget separately.

jgiannuzzi replied:
`dotnet pack` would be awesome, yes. I understand from @m4rs-mt that release-only builds are coming soon! 🥳

m4rs-mt replied:
Exactly; however, this "feature" requires some updates to the internal compiler architecture. We are currently working on it for v0.8.1.

adamreeve commented:
Thanks for the feedback, I've addressed all of the comments. I'll start adding automatic publishing of tagged releases to NuGet but think that might be best as a separate pull request.

adamreeve commented on Src/.config/dotnet-tools.json:
```diff
@@ -0,0 +1,12 @@
+  "isRoot": true,
+  "tools": {
+    "dotnet-t4-fix": {
+      "version": "2.3.0-preview-0000-gb1ee03d9bc",
```
This should be updated to use a stable release of dotnet-t4 once a new release is available. I ran into errors with the most recent stable release so had to switch to this pre-release version.

adamreeve commented:
I've added some additional changes to run the build and tests on Linux too, but have set up the problematic tests to be skipped when running on Linux for now.

jgiannuzzi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+        run: dotnet build --configuration=Release -p:TreatWarningsAsErrors=true Src
+      - name: CPU Tests
+        run: dotnet test --configuration=Release ./Src/ILGPU.Tests.CPU
+      - name: Set up NuGet
+        uses: nuget/setup-nuget@v1
+        with:
+          nuget-version: '5.x'
+      - name: Create NuGet package
+        run: nuget pack .nuget/ILGPU.nuspec
```
I think those 2 steps should run only on Windows

jgiannuzzi commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,40 @@
+  build-and-test:
+    strategy:
+      matrix:
+        os: [ubuntu-18.04, windows-latest]
```
I think we should add macos-latest in the test matrix

jgiannuzzi commented on Src/ILGPU.Tests/ArrayViews.cs:
```diff
@@ -381,7 +381,7 @@ public void ArrayViewGetVariableView(int length)
         }
 
-        [Theory]
+        [WindowsOnlyTheory("Fails to create type information for ArrayView on Linux")]
```
Comment should say "Unix" instead of "Linux" as this test also fails on macOS

jgiannuzzi commented on Src/ILGPU.Tests/StructureValues.cs:
```diff
@@ -124,7 +124,7 @@ public void StructureInterop<T>(T value)
         }
 
-        [Theory]
+        [WindowsOnlyTheory("Fails to create type information for ArrayView on Linux")]
```
Comment should say "Unix" instead of "Linux" as this test also fails on macOS

jgiannuzzi commented:
Looks like I'm a bit late to the party with this review 🤷🏻‍♂️
@m4rs-mt @adamreeve Should I create a PR to address my comments above?

adamreeve commented:
@jgiannuzzi I'm happy to create another PR to address those
